fideslib/models/client.py

```diff
@@ -0,0 +1,130 @@
+from __future__ import annotations
+
+import json
+from datetime import datetime
+from typing import Any
+
+from sqlalchemy import ARRAY, Column, ForeignKey, String
+from sqlalchemy.ext.declarative import declared_attr
+from sqlalchemy.orm import Session
+
+from fideslib.cryptography.cryptographic_util import (
+    generate_salt,
+    generate_secure_random_string,
+    hash_with_salt,
+)
+from fideslib.cryptography.schemas.jwt import (
+    JWE_ISSUED_AT,
+    JWE_PAYLOAD_CLIENT_ID,
+    JWE_PAYLOAD_SCOPES,
+)
+from fideslib.db.base_class import Base
+from fideslib.models.fides_user import FidesUser
+from fideslib.oauth.api.scope_registry import SCOPE_REGISTRY
+from fideslib.oauth.jwt import generate_jwe
+
+DEFAULT_SCOPES: list[str] = []
+
+
+class ClientDetail(Base):  # type: ignore
+    """The persisted details about a client in the system"""
+
+    @declared_attr
+    def __tablename__(cls) -> str:  # pylint: disable=E0213
+        return "client"
+
+    hashed_secret = Column(String, nullable=False)
+    salt = Column(String, nullable=False)
+    scopes = Column(ARRAY(String), nullable=False, default="{}")
+    fides_key = Column(String, index=True, unique=True, nullable=True)
+    user_id = Column(
+        String, ForeignKey(FidesUser.id_field_path), nullable=True, unique=True
+    )
+
+    @classmethod
+    def create_client_and_secret(
+        cls,
+        db: Session,
+        client_id_byte_length: int,
+        client_secret_byte_length: int,
+        *,
+        scopes: list[str] | None = None,
+        fides_key: str = None,
+        user_id: str = None,
+        encoding: str = "UTF-8",
+    ) -> tuple["ClientDetail", str]:
+        """Creates a ClientDetail and returns that along with the unhashed secret
+        so it can be returned to the user on create
+        """
+
+        client_id = generate_secure_random_string(client_id_byte_length)
+        secret = generate_secure_random_string(client_secret_byte_length)
+
+        if not scopes:
+            scopes = DEFAULT_SCOPES
+
+        salt = generate_salt()
+        hashed_secret = hash_with_salt(
+            secret.encode(encoding),
+            salt.encode(encoding),
+        )
+
+        client = super().create(
+            db,
+            data={
+                "id": client_id,
+                "salt": salt,
+                "hashed_secret": hashed_secret,
+                "scopes": scopes,
+                "fides_key": fides_key,
+                "user_id": user_id,
+            },
+        )
+        return client, secret
+
+    @classmethod
+    def get(
+        cls,
+        db: Session,
+        *,
+        table_id: Any,
+        root_client_id: str | None = None,
+        root_client_hash: tuple[str, str] | None = None,
+    ) -> ClientDetail | None:
+        """Fetch a database record via a table ID"""
+        if root_client_id and root_client_hash and table_id == root_client_id:
```

Comment:
I renamed the `id` parameter to `table_id` because `id` is a python built in function so redefining it could lead to bugs.

Reply:
to rephrase, I didn't understand why we needed both a root_client_id and a table_id here.

Reply:
Ah, ok. I was questioning things here so it's entirely possible I got it wrong. I'll remove the double.

Reply:
After looking to remove it now I remember why I did it, but it's still possible I got it wrong. The original:

```py
@classmethod
    def get(
        cls, db: Session, *, id: Any  # pylint: disable=W0622
    ) -> Optional["ClientDetail"]:
        """Fetch a database record via a table ID"""
        if id == config.security.OAUTH_ROOT_CLIENT_ID:
            return _get_root_client_detail()
        return super().get(db, id=id)
```

So my assumption was an id could be passed that is different than the `root_client_id` since there is a check to see if the `id` matches the `OAUTH_ROOT_CLIENT_ID`. With the change we are making does this mean that they will always be the same thing and should be returned unless it is `None`? So like this?

```py
@classmethod
def get(
    cls,
    db: Session,
    *,
    root_client_id: str | None = None,
    root_client_hash: tuple[str, str] | None = None,
) -> ClientDetail | None:
    """Fetch a database record via a table ID"""
    if root_client_id and root_client_hash :
            return _get_root_client_detail(root_client_id, root_client_hash)
    return super().get(db, id=id). # I'm also not sure about the id=id here
```

Reply:
OK I understand now, my confusion was just that I forgot you needed to pass in the `root_client_id` as an environment variable now, which is why you needed both `root_client_id` and the `table_id`.

The original purpose of this function was to return a pretend ClientDetail that's not actually in the database if it's the root client, otherwise, look up the ClientDetail by its id. An id can definitely be passed that is different than the root_client_id.  I would typically just use the root client in practice to create a separate client.

So your first version was closer, I would just rename `table_id` to `client_id` for clarity, and make sure you're using the passed in `client_id` to fetch it from the database in the second return.  

```python
 @classmethod
    def get(
        cls,
        db: Session,
        *,
        client_id: Any,
        root_client_id: str | None = None,
        root_client_hash: tuple[str, str] | None = None,
    ) -> ClientDetail | None:
        """Fetch a database record via a client_id"""
        if root_client_id and root_client_hash and client_id == root_client_id:
            return _get_root_client_detail(root_client_id, root_client_hash)
        return super().get(db, id=client_id)
```